SRC/dlassq.f90

```diff
@@ -200,13 +191,43 @@ subroutine DLASSQ( n, x, incx, scale, sumsq )
    if( sumsq > zero ) then
       ax = scale*sqrt( sumsq )
       if (ax > tbig) then
-!        We assume scale >= sqrt( TINY*EPS ) / sbig
-         abig = abig + (scale*sbig)**2 * sumsq
+         if (scale > one) then
+            scale = scale * sbig  ! sbig < scale <= sbig * max
+            if (scale > sqrtmin) then
+               ! sqrtmin < scale < sqrtmax, so it is safe to square scale
+               abig = abig + (scale * scale) * sumsq
+            else
+               ! Do not square scale, as it may underflow
+               abig = abig + scale * (scale * sumsq)
+            end if
```

Comment:
Is is possible to remove the if branch altogether and always take the accumulation order of the else branch, ie always accumulate `scale * (scale * sumsq)` ?

Reply:
I think so. Thanks a lot! I will try to write some notes with the upper and lower bounds on those quantities. I will provide the improvement in the next commit.

Reply:
Let me write the reasoning behind the changes. It simpler than I first thought.

Here is the new code: 

```f90
   if( sumsq > zero ) then
      ax = scale*sqrt( sumsq )
      if (ax > tbig) then
         if (scale > one) then
            scale = scale * sbig
            abig = abig + scale * (scale * sumsq)
         else
            ! sumsq > tbig^2 => (sbig * (sbig * sumsq)) is representable
            abig = abig + scale * (scale * (sbig * (sbig * sumsq)))
         end if
      else if (ax < tsml) then
         if (notbig) then
            if (scale < one) then
               scale = scale * ssml
               asml = asml + scale * (scale * sumsq)
            else
               ! sumsq < tsml^2 => (ssml * (ssml * sumsq)) is representable
               asml = asml + scale * (scale * (ssml * (ssml * sumsq)))
            end if
         end if
      else
         amed = amed + scale * (scale * sumsq)
      end if
   end if
```

Cases:
1. `ax > tbig` and `scale > one`:
    - we can safely compute the update `scale_updated = scale * sbig` because `sbig < one` and `scale > one`.
    - `scale_updated * sumsq` does not underflow. In fact, if `scale_updated * sumsq` were to underflow, then we would know that `scale_updated < one` and `sumsq < one`. In this case, `scale_updated * scale_updated * sumsq` would also underflow independently on how we organize the operations. This would contradict the hypothesis that `(ax * sbig) * (ax * sbig) = scale_updated * scale_updated * sumsq` is representable. So, indeed, `scale_updated * sumsq` does not underflow.
    - `scale_updated * sumsq` does not overflow using an argument similar to the above.
2. `ax > tbig` and `scale <= one`:
    - In this case, we do not know if `scale * sbig` is representable. However, we know that `sumsq > tbig * tbig` and, thus, `(sbig * (sbig * sumsq))` is representable.
3. `ax < tsml` and `scale < one`: Reasoning is analogous to (1).
4. `ax < tsml` and `scale >= one`: Reasoning is analogous to (2).
5. `tsml <= ax <= tbig`: Reasoning is analogous to the second bullet in (1).